metadata/src/main/java/org/apache/kafka/controller/MetaLogManager.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ApiMessageAndVersion;
+
+/**
+ * The MetaLogManager handles storing metadata and electing leaders.
+ */
+public interface MetaLogManager extends AutoCloseable {
+    /**
+     * Listeners receive notifications from the MetaLogManager.
+     */
+    interface Listener {
+        /**
+         * Called when the MetaLogManager commits a message.
+         *
+         * @param epoch         The controller epoch of the message.
+         * @param index         The index of the message within the controller epoch.
+         * @param message       The message.
+         */
+        void commit(long epoch, long index, ApiMessage message);
```

Comment:
Wondering why we don't pass in the full `ApiMessageAndVersion` here.  I guess because the version is irrelevant at this point, but wanted to raise the question just in case.

Reply:
Yes, the version is irrelevant here.  We don't want the code to be littered with if version > foo checks, since new versions should have reasonable defaults for any added fields so that old versions work fine.  Or if that can't be done, we should create new message types for the new messages.